custom/enikshay/tasks.py

```diff
@@ -0,0 +1,180 @@
+from corehq.util.soft_assert import soft_assert
+from dimagi.utils.decorators.memoized import memoized
+
+from .case_utils import get_adherence_cases_between_dates_from_episode, CASE_TYPE_ADHERENCE, CASE_TYPE_EPISODE
+
+
+DOSE_TAKEN_INDICATORS = [
+    'directly_observed_dose',
+    'unobserved_dose',
+    'self_administered_dose',
+]
+DAILY_SCHEDULE_FIXTURE_NAME = 'adherence_schedules'
+DAILY_SCHEDULE_ID = 'schedule_daily'
+SCHEDULE_ID_FIXTURE = 'id'
+
+
+@periodic_task(
+    run_every=crontab(day_of_week=[1], hour=0, minute=0),  # every Monday
+    queue=getattr(settings, 'CELERY_PERIODIC_QUEUE', 'celery')
+)
+def enikshay_adherence_task():
+    domains = toggles.UATBC_ADHERENCE_TASK.get_enabled_domains()
+    for domain in domains:
+        updater = PeriodicCaseUpdater(domain)
+        updater.run()
+
+
+class PeriodicCaseUpdater(object):
+
+    def __init__(self, domain):
+        self.domain = domain
+        self.purge_date = pytz.UTC.localize(datetime.datetime.today() - datetime.timedelta(days=60))
+
+    def run(self):
+        for episode in self._get_open_episode_cases():
+            update = EpisodeUpdate(episode, self)
+            case_block = update.case_block()
+            if case_block:
+                submit_case_blocks(
+                    [ElementTree.tostring(case_block.as_xml())],
+                    self.domain
+                )
+
+    def _get_open_episode_cases(self):
+        case_accessor = CaseAccessors(self.domain)
+        case_ids = case_accessor.get_open_case_ids_in_domain_by_type(CASE_TYPE_EPISODE)
+        return case_accessor.iter_cases(case_ids)
+
+    @memoized
+    def get_doses_data(self):
+        # return 'doses_per_week' by 'schedule_id' from the Fixture data
+        fixtures = FixtureDataItem.get_indexed_items(self.domain, DAILY_SCHEDULE_FIXTURE_NAME, SCHEDULE_ID_FIXTURE)
+        return dict((k, int(fixture['doses_per_week'])) for k, fixture in fixtures.items())
+
+
+def index_by_adherence_date(adherence_cases):
+    """
+    index by day of adherence_date datetime
+    """
+    by_date = defaultdict(list)
+    for case in adherence_cases:
+        adherence_date = parse_datetime(case.dynamic_case_properties().get('adherence_date')).date()
+        by_date[adherence_date].append(case)
+    return by_date
+
+
+class EpisodeUpdate(object):
+
+    def __init__(self, episode_case, case_updater):
+        self.episode = episode_case
+        self.case_updater = case_updater
+
+    def get_property(self, property):
+        return self.episode.dynamic_case_properties().get(property)
+
+    def get_latest_adherence_case_for_episode(self):
+        """
+        return open case of type 'adherence' reversed-indexed to episode and
+            with latest 'adherence_date' property
+        """
+        case_accessor = CaseAccessors(self.case_updater.domain)
+        indexed_cases = case_accessor.get_reverse_indexed_cases([self.episode.case_id])
```

Comment:
it's pretty tempting to try and push this logic into elasticsearch. as-is it seems like this will have to load every single adherence case for every single episode in the entire domain every time it runs right? I think that is going to be pretty untenable long term. could we possibly use the case search index for this?